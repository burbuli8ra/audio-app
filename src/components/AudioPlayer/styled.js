import styled from 'styled-components';

const AudioPlayer = styled.div`
  background: #ffffff;
  box-shadow: 0 0 9px 0 rgba(140, 140, 158, 0.25);
  padding: 20px 20px 32px 20px;
  position: sticky;
  bottom: 0;
`;
AudioPlayer.displayName = 'AudioPlayer';

const Artist = styled.div`
  color: #404040;
  font-size: 14px;
`;
Artist.displayName = 'Artist';

const Notification = styled.div`
  color: #28aa90;
  font-size: 14px;
  font-style: italic;
`;
Notification.displayName = 'Notification';

const Title = styled.div`
  color: #28aa90;
  font-size: 18px;
  font-weight: bold;
`;
Title.displayName = 'Title';

const Info = styled.div`
  display: flex;
  align-items: center;
  flex-direction: column;
  justify-content: center;
  min-height: 45px;
  padding-bottom: 8px;
`;
Info.displayName = 'Info';

const ControlButton = styled.div`
  svg {
    color: ${({ isDisabled }) => isDisabled ? '#d3d3d3' : '#28aa90'};
    cursor: ${({ isDisabled }) => isDisabled ? 'not-allowed' : 'pointer'};
    font-size: ${({ isMain }) => isMain ? '64px' : '42px'};
  }

  &:hover {
    svg {
      color: ${({ isDisabled }) => isDisabled ? '#d3d3d3' : '#3cd1b3'};
    }
  }
`;
ControlButton.displayName = 'ControlButton';

const Controls = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;
Controls.displayName = 'Controls';

const Time = styled.div`
  color: ${({ isDisabled }) => isDisabled ? '#d3d3d3' : '#404040'};
  display: inline-flex;
`;
Time.displayName = 'Time';

const ProgressBar = styled.div.attrs(
  ({ currentProgress }) => ({
    style: {
      backgroundImage: `linear-gradient(to right, #28aa90 ${currentProgress }%, #d3d3d3 0)`
    }
  }))`
  border-radius: 5px;
  height: 10px;
  flex-grow: 1;
  margin: 0 20px;
`;
ProgressBar.displayName = 'ProgressBar';

const Progress = styled.div`
  color: #404040;
  display: flex;
  align-items: center;
  font-size: 14px;
`;
Progress.displayName = 'Progress';

export default {
  Artist,
  AudioPlayer,
  ControlButton,
  Controls,
  Info,
  Notification,
  Progress,
  ProgressBar,
  Time,
  Title
};