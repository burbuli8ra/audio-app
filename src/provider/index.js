import React, { createContext, useContext, useReducer } from 'react';
import activeArtistReducer from './reducers/activeArtistReducer';
import activeAudioIndexReducer from './reducers/activeAudioIndexReducer';
import activeTrackListReducer from './reducers/activeTrackListReducer';

export const AppContext = createContext();

export const initialState = {
  activeArtist: null,
  activeAudioIndex: null,
  activeTrackList: []
};

export const mainReducer = (state, action) => ({
  ...state,
  activeArtist: activeArtistReducer(state.activeArtist, action),
  activeAudioIndex: activeAudioIndexReducer(state.activeAudioIndex, action),
  activeTrackList: activeTrackListReducer(state.activeTrackList, action)
});

export const StateProvider = ({ reducer, initialState, children }) => (
  <AppContext.Provider value={useReducer(reducer, initialState)}>
    {children}
  </AppContext.Provider>
);

export const useAppContext = () => useContext(AppContext);