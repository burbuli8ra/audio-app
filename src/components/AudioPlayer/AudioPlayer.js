import React, { useEffect, useRef, useState } from 'react';
import { useAppContext } from '../../provider';
import Styled from './styled'
import {
  PlayCircleOutlineRounded,
  PauseCircleOutlineRounded,
  SkipNextRounded,
  SkipPreviousRounded
} from '@material-ui/icons';

const AudioPlayer = () => {
  const audioRef = useRef();

  const [isPlaying, setIsPlaying] = useState(false);
  const [currentTime, setCurrentTime] = useState();
  const [duration, setDuration] = useState();

  const [{ activeArtist, activeAudioIndex, activeTrackList }, dispatch] = useAppContext();
  
  const currentProgress = currentTime / duration * 100 || 0;
  const activeTrack = ((activeAudioIndex || activeAudioIndex === 0) && activeTrackList)
    ? activeTrackList[activeAudioIndex]
    : {};
  const isEmptySource = !Boolean(activeTrack.url);
  const isPrevDisabled = Boolean(!activeAudioIndex);
  const isNextDisabled = typeof(activeAudioIndex) !== 'number'
    || activeTrackList.length - 1 === activeAudioIndex;

  const formatTime = (seconds = 0) => {
    // Create date object with milliseconds
    const date = new Date(seconds * 1000);
    
    // In case of hours get hours, minutes and seconds
    if (date.getUTCHours()) {
      return date.toISOString().substr(11, 8);
    }

    // Get minutes and seconds
    return date.toISOString().substr(14, 5);
  };

  const handlePlayPause = () => {
    if (!isEmptySource) {
      setIsPlaying(!isPlaying);
    }
  };

  const handlePrevClick = () => {
    if (!isPrevDisabled) {
      dispatch({
        type: 'SET_ACTIVE_INDEX',
        payload: activeAudioIndex - 1
      });
    }
  };

  const handleNextClick = () => {
    if (!isNextDisabled) {
      dispatch({
        type: 'SET_ACTIVE_INDEX',
        payload: activeAudioIndex + 1
      });
    }
  };

  useEffect(() => {
    const audio = audioRef.current;

    if (activeTrack.url) {
      setIsPlaying(false);
      audio.load();
    }
  }, [activeTrack.url, activeTrack.title]);

  useEffect(() => {
    const audio = audioRef.current;

    const setAudioData = () => {
      setDuration(audio.duration);
      setAudioTime();
      setIsPlaying(true);
    };

    const setAudioTime = () => {
      setCurrentTime(audio.currentTime);
    };

    const resetAudioTime = () => {
      setIsPlaying(false);
      audio.currentTime = 0;
      setCurrentTime(0);
    };

    const setNextAudio = () => {
      dispatch({
        type: 'SET_ACTIVE_INDEX',
        payload: activeAudioIndex + 1
      });
    };

    const resetAudioPlayer = () => {
      dispatch({ type: 'RESET_ACTIVE_INDEX' });
      dispatch({ type: 'RESET_ACTIVE_LIST' });
      dispatch({ type: 'RESET_ACTIVE_ARTIST' });
    };

    const handleAudioEnd = () => {
      resetAudioTime();
      
      // Reset audio player on last song of list
      if (activeAudioIndex === activeTrackList.length - 1) {
        resetAudioPlayer();
        return;
      }

      setNextAudio();
    };

    // Add event listeners
    audio.addEventListener('loadeddata', setAudioData);
    audio.addEventListener('timeupdate', setAudioTime);
    audio.addEventListener('ended', handleAudioEnd);

    // Cleanup
    return () => {
      audio.removeEventListener('loadeddata', setAudioData);
      audio.removeEventListener('timeupdate', setAudioTime);
      audio.removeEventListener('ended', handleAudioEnd);
    }
  }, [activeAudioIndex, activeTrackList.length, currentTime, dispatch]);


  useEffect(() => {
    const audio = audioRef.current;

    isPlaying ? audio.play() : audio.pause();
  }, [isPlaying]);

  return (
    <Styled.AudioPlayer>
      <audio
        data-testid="audio"
        ref={audioRef}
      >
        <source src={activeTrack.url} type={activeTrack.type} />
        Your browser does not support the audio element.
      </audio>
      <Styled.Info>
        {activeTrack.url ? (
          <>
            <Styled.Title data-testid="audio-title">
              {activeTrack.title || 'Unknown song'}
            </Styled.Title>
            <Styled.Artist data-testid="audio-artist">
              {activeArtist || 'Unknown artist'}
            </Styled.Artist>
          </>
        ) : (
          <Styled.Notification data-testid="audio-notification">
            You have not selected any song yet!
          </Styled.Notification>
        )}
      </Styled.Info>
      <Styled.Controls>
        <Styled.ControlButton
          data-testid="audio-control-prev"
          isDisabled={isPrevDisabled}
          onClick={handlePrevClick}
        >
          <SkipPreviousRounded />
        </Styled.ControlButton>
        <Styled.ControlButton
          data-testid="audio-control-play-pause"
          isDisabled={isEmptySource}
          isMain
          onClick={handlePlayPause}
        >
          {isPlaying
            ? <PauseCircleOutlineRounded data-testid="audio-pause-icon" />
            : <PlayCircleOutlineRounded data-testid="audio-play-icon" />
          }
        </Styled.ControlButton>
        <Styled.ControlButton
          data-testid="audio-control-next"
          isDisabled={isNextDisabled}
          onClick={handleNextClick}
        >
          <SkipNextRounded />
        </Styled.ControlButton>
      </Styled.Controls>
      <Styled.Progress
        currentTime={currentTime}
        totalTime={duration}
      >
        <Styled.Time isDisabled={isEmptySource}>
          {formatTime(currentTime)}
        </Styled.Time>
        <Styled.ProgressBar currentProgress={currentProgress} />
        <Styled.Time isDisabled={isEmptySource}>
          {formatTime(duration)}
        </Styled.Time>
      </Styled.Progress>
    </Styled.AudioPlayer>
  );
};

export default AudioPlayer;