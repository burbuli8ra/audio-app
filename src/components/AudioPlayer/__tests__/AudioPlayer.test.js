import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { StateProvider, initialState, mainReducer } from '../../../provider';
import albumInfoMock from '../../../__mocks__/album.json';
import AudioPlayer from '../AudioPlayer';

describe('AudioPlayer component', () => {
  let wrapper;

  const setup = (state = initialState) => (
    <StateProvider reducer={mainReducer} initialState={state}>
      <AudioPlayer />
    </StateProvider>
  );

  beforeEach(() => {
    window.HTMLMediaElement.prototype.load = () => {};
    window.HTMLMediaElement.prototype.play = () => {};
    window.HTMLMediaElement.prototype.pause  = () => {};
  });

  test('should match snapshot', () => {
    wrapper = render(setup());
    expect(wrapper).toMatchSnapshot();
  });

  test('should show song title and artist', () => {
    const initialStateMock = {
      activeArtist: albumInfoMock[1].artist,
      activeAudioIndex: 2,
      activeTrackList: albumInfoMock[1].tracks
    };

    wrapper = render(setup(initialStateMock));
    const { getByTestId, queryByTestId } = wrapper;

    expect(getByTestId('audio-title')).toBeInTheDocument();
    expect(getByTestId('audio-artist')).toBeInTheDocument();
    expect(queryByTestId('audio-notification')).toBeNull();
  });

  test('should play previous song', () => {
    const initialStateMock = {
      activeArtist: albumInfoMock[1].artist,
      activeAudioIndex: 2,
      activeTrackList: albumInfoMock[1].tracks
    };

    wrapper = render(setup(initialStateMock));
    const { getByTestId } = wrapper;

    expect(getByTestId('audio-title').textContent).toBe(albumInfoMock[1].tracks[2].title);
    fireEvent.click(getByTestId('audio-control-prev'));
    expect(getByTestId('audio-title').textContent).toBe(albumInfoMock[1].tracks[1].title);
  });

  test('should pause song', () => {
    const initialStateMock = {
      activeArtist: albumInfoMock[1].artist,
      activeAudioIndex: 2,
      activeTrackList: albumInfoMock[1].tracks
    };

    wrapper = render(setup(initialStateMock));
    const { getByTestId } = wrapper;

    expect(getByTestId('audio-title').textContent).toBe(albumInfoMock[1].tracks[2].title);
    fireEvent.click(getByTestId('audio-control-next'));
    expect(getByTestId('audio-title').textContent).toBe(albumInfoMock[1].tracks[3].title);
  });

  test('should play next song', () => {
    const initialStateMock = {
      activeArtist: albumInfoMock[1].artist,
      activeAudioIndex: 2,
      activeTrackList: albumInfoMock[1].tracks
    };

    wrapper = render(setup(initialStateMock));
    const { getByTestId, queryByTestId } = wrapper;

    expect(getByTestId('audio-play-icon')).toBeInTheDocument();
    expect(queryByTestId('audio-pause-icon')).toBeNull();
    fireEvent.click(getByTestId('audio-control-play-pause'));
    expect(getByTestId('audio-pause-icon')).toBeInTheDocument();
    expect(queryByTestId('audio-play-icon')).toBeNull();
  });
});