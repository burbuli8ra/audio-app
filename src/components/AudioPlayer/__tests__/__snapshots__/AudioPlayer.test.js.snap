// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioPlayer component should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-AxjAm jkZftL"
      >
        <audio
          data-testid="audio"
        >
          <source />
          Your browser does not support the audio element.
        </audio>
        <div
          class="sc-AxhUy dMGOsD"
        >
          <div
            class="sc-AxiKw cxefkL"
            data-testid="audio-notification"
          >
            You have not selected any song yet!
          </div>
        </div>
        <div
          class="sc-AxheI JGUfr"
        >
          <div
            class="sc-AxgMl lnWYQu"
            data-testid="audio-control-prev"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 6c.55 0 1 .45 1 1v10c0 .55-.45 1-1 1s-1-.45-1-1V7c0-.55.45-1 1-1zm3.66 6.82l5.77 4.07c.66.47 1.58-.01 1.58-.82V7.93c0-.81-.91-1.28-1.58-.82l-5.77 4.07c-.57.4-.57 1.24 0 1.64z"
              />
            </svg>
          </div>
          <div
            class="sc-AxgMl gBRePK"
            data-testid="audio-control-play-pause"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              data-testid="audio-play-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10.8 15.9l4.67-3.5c.27-.2.27-.6 0-.8L10.8 8.1c-.33-.25-.8-.01-.8.4v7c0 .41.47.65.8.4zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </div>
          <div
            class="sc-AxgMl lnWYQu"
            data-testid="audio-control-next"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.58 16.89l5.77-4.07c.56-.4.56-1.24 0-1.63L7.58 7.11C6.91 6.65 6 7.12 6 7.93v8.14c0 .81.91 1.28 1.58.82zM16 7v10c0 .55.45 1 1 1s1-.45 1-1V7c0-.55-.45-1-1-1s-1 .45-1 1z"
              />
            </svg>
          </div>
        </div>
        <div
          class="sc-fzozJi cCmWGe"
        >
          <div
            class="sc-Axmtr dILVTF"
          >
            00:00
          </div>
          <div
            class="sc-AxmLO bptprH"
          />
          <div
            class="sc-Axmtr dILVTF"
          >
            00:00
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-AxjAm jkZftL"
    >
      <audio
        data-testid="audio"
      >
        <source />
        Your browser does not support the audio element.
      </audio>
      <div
        class="sc-AxhUy dMGOsD"
      >
        <div
          class="sc-AxiKw cxefkL"
          data-testid="audio-notification"
        >
          You have not selected any song yet!
        </div>
      </div>
      <div
        class="sc-AxheI JGUfr"
      >
        <div
          class="sc-AxgMl lnWYQu"
          data-testid="audio-control-prev"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 6c.55 0 1 .45 1 1v10c0 .55-.45 1-1 1s-1-.45-1-1V7c0-.55.45-1 1-1zm3.66 6.82l5.77 4.07c.66.47 1.58-.01 1.58-.82V7.93c0-.81-.91-1.28-1.58-.82l-5.77 4.07c-.57.4-.57 1.24 0 1.64z"
            />
          </svg>
        </div>
        <div
          class="sc-AxgMl gBRePK"
          data-testid="audio-control-play-pause"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            data-testid="audio-play-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10.8 15.9l4.67-3.5c.27-.2.27-.6 0-.8L10.8 8.1c-.33-.25-.8-.01-.8.4v7c0 .41.47.65.8.4zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </div>
        <div
          class="sc-AxgMl lnWYQu"
          data-testid="audio-control-next"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.58 16.89l5.77-4.07c.56-.4.56-1.24 0-1.63L7.58 7.11C6.91 6.65 6 7.12 6 7.93v8.14c0 .81.91 1.28 1.58.82zM16 7v10c0 .55.45 1 1 1s1-.45 1-1V7c0-.55-.45-1-1-1s-1 .45-1 1z"
            />
          </svg>
        </div>
      </div>
      <div
        class="sc-fzozJi cCmWGe"
      >
        <div
          class="sc-Axmtr dILVTF"
        >
          00:00
        </div>
        <div
          class="sc-AxmLO bptprH"
        />
        <div
          class="sc-Axmtr dILVTF"
        >
          00:00
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
